### DISCUSSION TITLE
Sum types having members with no attributes

### DISCUSSION THREAD

**Discussion by drhagen** [eclipse-langium/langium#1348] [Q&A] [answered]
What is the appropriate way to model sum types where one or more of the members has no payload. For example, the naive grammar for the `Option` type would be:

```langium
grammar HelloWorld

entry Option:
    Some | None;

Some:
    'Some' '(' value = ID ')';

None:
    'None' '(' ')';

hidden terminal WS: /\s+/;
terminal ID: /[a-z]+/;
```

But this does not work because `None` has no attributes, which triggers an error: `Data type rules cannot infer a type.`

How do I tell Langium that `None` should be inferred to be an interface like:

```typescript
export interface None extends AstNode {
    readonly $type: 'None';
}
```

**Comment by msujew** [comment, marked as answer]
Hey @drhagen,

you can just [invoke an action](https://langium.org/docs/grammar-language/#simple-actions) to force Langium to instantiate an object (and therefore create a corresponding type in the `ast.ts`):

```
None:
    {infer None} 'None' '(' ')';
```
